Expose FS translation system to plugins (#2578)
withastro/starlight: docs/src/content/docs/reference/plugins.md

## `hooks`

Hooks are functions which Starlight calls to run plugin code at specific times.

To get the type of a hook's arguments, use the `HookParameters` utility type and pass in the hook name.

> In the following example, the `configSetup` function `options` parameter is typed to match the parameters of the `config:setup` hook:

Reviewer: Suggesting a small simplification here to help make the sentence a little bit clearer:

```suggestion
In the following example, the `options` parameter is typed to match the arguments passed to the `config:setup` hook:
```

```ts
import type { HookParameters } from '@astrojs/starlight/types';

function configSetup(options: HookParameters['config:setup']) {
  options.useTranslations('en');
}
```

### `i18n:setup`

> Plugin internationalization setup function called when Starlight is initialized (during the [`astro:config:setup`](https://docs.astro.build/en/reference/integrations-reference/#astroconfigsetup) integration hook).

Reviewer: What do you think about omitting the Astro hook detail? Part of me feels like it might be superfluous information that distracts from the focus of this section, but I also get that it could be useful information.

```suggestion
Plugin internationalization setup function called when Starlight is initialized.
```

> The `i18n:setup` hook can be used to inject translations strings for the plugin in various locales.

Reviewer: Small tweak suggestion:

```suggestion
The `i18n:setup` hook can be used to inject translation strings so a plugin can support different locales.
```

> The `i18n:setup` hook can be used to inject translations strings for the plugin in various locales.
> These translations will be available in the [`config:setup`](#configsetup) hook and plugin UI.

Reviewer: Suggesting a small tweak here. “plugin UI” sounds a bit like some kind of UI for managing a plugin somehow to me, so wanted to try to emphasise “components” to help with understanding that this is in your website’s UI and how to access them.

```suggestion
The `i18n:setup` hook can be used to inject translation strings so a plugin can support different locales.
These translations will be available via [`useTranslations()`](#usetranslations) in the `config:setup` hook and in UI components via [`Astro.locals.t()`](/guides/i18n/#using-ui-translations).
```

> This hook is called with the following options:

Reviewer: I think it might be worth repeating the hook name here rather than just referring to it with “this“:

```suggestion
The `i18n:setup` hook is called with the following options:
```

[long-process-plugin] Starting long process…
```

#### `useTranslations`

**type:** `(lang: string) => I18nT`

A callback function to generate a utility function to access UI strings for a given language.
To learn more about the generated utility function and all the APIs available, see the [“Using UI translations”](/guides/i18n/#using-ui-translations) guide.

```ts {6}
// plugin.ts
export default {
  name: 'plugin-use-translations',
  hooks: {
    setup({ useTranslations, logger }) {
      const t = useTranslations('en');
      logger.info(t('builtWithStarlight.label'));
    },
  },
};
```

The example above will log a message that includes a built-in UI string for the English language:

```shell
[plugin-use-translations] Built with Starlight
```

#### `pathToLang`

**type:** `(path: string) => string`

> A callback function to get the language for a given full file path.

Reviewer: Aligning this with the previous suggestion, which I also like because it makes clear the arguments to pass as well as the purpose.

```suggestion
Call `pathToLang()` with a file path to get the language for that file.
```

One question here though: in general these docs always say “full” file path. What happens if I do `pathToLang('src/content/docs/en/foo.md')`? Or even `pathToLang('en/foo.md')`? Is it dependent on being an absolute path?

Author: I can add a few test cases for these, but they all fallback to the default locale.

> This can be particularly useful in [remark or rehype plugins](https://docs.astro.build/en/guides/markdown-content/#markdown-plugins) that a Starlight plugin may use to process Markdown or MDX files.
> The [virtual file format](https://github.com/vfile/vfile) used by such plugins includes the [full path](https://github.com/vfile/vfile#filepath) of the file being processed, which can be used with `pathToLang` to determine the language of the file.
> The returned language can be used with the [`useTranslations`](#usetranslations) helper to get UI strings for that language.

Reviewer: Suggesting some tiny tweaks here:
- removed the “Starlight plugin” part as it helps simplify the sentence and hopefully it’s obvious on this page we’re referring to plugins
- added `()` to the mention of the functions like we usually do

```suggestion
This can be particularly useful when adding [remark or rehype plugins](https://docs.astro.build/en/guides/markdown-content/#markdown-plugins) to process Markdown or MDX files.
The [virtual file format](https://github.com/vfile/vfile) used by these plugins includes the [full path](https://github.com/vfile/vfile#filepath) of the file being processed, which can be used with `pathToLang()` to determine the language of the file.
The returned language can be used with the [`useTranslations()`](#usetranslations) helper to get UI strings for that language.
```
